### PR TITLE
NetSim: Leading-edge refresh delay to better coalesce grouped invalidations

### PR DESCRIPTION
**Pull request by islemaster**
Before, we only used `_.throttle()` with leading and (conditional) trailing events to limit our request rate.  Our table change notifications often show up in little groups, and we would get two refresh events for each of these groups, like this:

```
onTableChanged    :   || |                     | |
throttled refresh :   |---throttled--->|       |---throttled--->|
```

When in fact, we'd rather have a short delay before the request is triggered (a `debounce`) so that the whole group of notifications only causes one refresh, not two.

```
onTableChanged    :   || |                     | |
debounced refresh :   ----debounce->|          ---debounce->|
```

But, we don't _just_ want a debounce because in high-traffic times you'd get perpetually delayed, like this:

```
onTableChanged    :   || |        |     |      | |
debounced refresh :   ----------------------------debounce->|
```

And we also want that initial delay to be shorter than our regular delay between requests, so that we're responsive when things are idle.

What this PR ends up at is our original throttle function (1000ms with leading and trailing) wrapped in a 250ms debounce with a maxWait of 250ms (which, if you're keeping track, is the same as a 250ms throttle with only trailing events.  I went with debounce because we might want to tune that maxWait to something different than the base delay).  Something like this, generating two requests instead of four in low traffic:

```
onTableChanged    :   || |                     | |
debounced         :   -250>|                   -250>|
compound refresh  :        |--------------1000->    |--------------1000->
```

And like this in higher traffic, generating three events at one-second intervals, but starting only 250ms after the first invalidation.

```
onTableChanged    :   || |        |     |      | |
debounced         :   -250>|      -250>|-250>| -250>|
compound refresh  :        |--------------1000->|--------------1000->|
```

Another feature not demonstrated above is that we add a small random additional delay to the debounce portion of the process, so that clients responding to the same invalidation events won't all hit the server at the same time.

### CODE REVIEW

**Comment by islemaster** [comment]
@philbogle Please review.

The 'jitter' isn't randomized all the time right now, it's set once when we create the compound-throttled function and then persists for that client.  I'm having trouble figuring out when I can safely re-randomize it; as currently implemented, that would mean recreating the debounce function at least, which would throw away some of the embedded information that's critical to doing the debouncing.  Do I randomize the jitter amount every time we actually refresh?

File: apps/src/netsim/NetSimTable.js [hunk 1]
```diff
@@ -119,6 +148,38 @@ var NetSimTable = module.exports = function (channel, shardID, tableName, option
   this.lastRefreshTime_ = 0;
 
   /**
+   * Minimum time (in ms) to wait after an invalidation event before attempting
+   * to trigger a refresh request.  This produces a window in which clustered
+   * invalidations can be captured and coalesced together.
+   * @private {number}
+   */
+  this.minimumDelayBeforeRefresh_ = DEFAULT_MINIMUM_DELAY_BEFORE_REFRESH_MS;
+  if (options && options.minimumDelayBeforeRefresh !== undefined) {
```
Comment (philbogle) [inline apps/src/netsim/NetSimTable.js]: Is "!== undefined" what we want here?  It seems like if it was null you wouldn't want to set it either.

File: apps/src/netsim/NetSimTable.js [hunk 2]
```diff
@@ -18,6 +18,7 @@
 var _ = require('../utils').getLodash();
 var ObservableEvent = require('../ObservableEvent');
 var NetSimApi = require('./NetSimApi');
+var netsimGlobals = require('./netsimGlobals');
```
Comment (philbogle) [inline apps/src/netsim/NetSimTable.js]: I think this should be capitalized for consistency with the other "modules" you are are requiring.

**Comment by philbogle** [comment]
I wouldn't worry too much about the fact that the jitter isn't getting re-randomized every time; the simplicity benefits you get out of using the predefined debounce function are valuable and in practice I don't think it will change things greatly.

File: apps/src/netsim/NetSimTable.js [hunk 3]
```diff
@@ -189,14 +249,16 @@ NetSimTable.prototype.refresh = function (callback) {
 /**
  * Generate throttled refresh function which will generate actual server
  * requests at the maximum given rate no matter how fast it is called. This
- * allows us to coalesce refresh events and reduce server load.
- * @param {number} waitMs - Minimum time (in milliseconds) to wait between
- *        refresh requests to the server.
+ * allows us to coalesce refreshAll events and reduce server load.
  * @returns {function}
```
Comment (philbogle) [inline apps/src/netsim/NetSimTable.js]: Could we follow closure compiler conventions for documenting function arguments? (https://developers.google.com/closure/compiler/docs/js-for-compiler?csw=1#types)

If the returned function takes no parameters and returns nothing interesting, you could just say "{function()}".

**Comment by philbogle** [comment]
LGTM (after small comments)

File: apps/src/netsim/NetSimTable.js (hunk 1 above)
Comment (islemaster) [inline apps/src/netsim/NetSimTable.js]: `undefined` is the value if the caller didn't set `minimumDelayBeforeRefresh` at all.  If this comes in `null` then the caller explicitly passed `minimumDelayBeforeRefresh: null` and we should probably throw `TypeError`.  I'll add a check for that.

**Pull request merged**